### PR TITLE
Enable rubocop on more of the analytics

### PR DESCRIPTION
**Pull request by ldodds**
I'm not sure that Rubocop was properly configured or properly running across the analytics. I've found a range of issues which haven't been caught before.

This PR:

* Updates Rubocop to latest version
* Updates the configuration so that its running across all the specs and all of the `app/classes` and `app/services` folders (initially)
* Runs rubocop to autocorrect the majority of errors
* Generates `.rubocop_todo.yml` by running `rubocop --auto-gen-config --exclude-limit 1000`
* Fixes up all of the basic issues with the specs, e.g. context naming
* Fixes up a number of minor issues across the classes and services
* Adds a custom naming convention to the Rubocop configuration to accept method and variable naming like `foo_£` which we use extensively across the code base

Follow on PRs will look at updating to cover `app/models` and fixing some of the items in the todo list. Goal is to resolve those before making progress on the `lib` folder which will likely need some significant rework. So will focus on removing old/unused code there first.

### CODE REVIEW

**Comment by iant** [comment]
hi @ldodds it may be worth also adding a rubocop github action step by adding something like this to https://github.com/Energy-Sparks/energy-sparks_analytics/blob/master/.github/workflows/ruby.yml (I normally have rubocop enabled in overcommit but have found it useful to have as an additional check in github before merging (along with [brakeman](https://github.com/presidentbeef/brakeman) and [bundler-audit](https://github.com/rubysec/bundler-audit)))
```
  rubocop:
    runs-on: ubuntu-latest
    steps:
      - name: Checkout code
        uses: actions/checkout@v3
      - name: Install Ruby and gems
        uses: ruby/setup-ruby@0a29871fe2b0200a17a4497bae54fe5df0d973aa # v1.115.3
        with:
          bundler-cache: true
      - name: Lint Ruby files (rubocop)
        run: bundle exec rubocop --parallel 
```
(I left a similar comment on https://github.com/Energy-Sparks/energy-sparks/pull/3141)

File: app/services/costs/accounting_costs_service.rb [hunk 1]
```diff
@@ -81,30 +84,29 @@ def dates_for_period(period)
...
         [last_year_end_date - DAYSINYEAR + 1, last_year_end_date]
       else
-        raise "Invalid year"
+        raise 'Invalid year'
       end
     end
 
     def has_full_previous_years_worth_of_data?
-      previous_year_start_date, previous_year_end_date = dates_for_period(:last_year)
+      previous_year_start_date, = dates_for_period(:last_year)
```
Comment (urbanwide) [inline app/services/costs/accounting_costs_service.rb:92]: This looks a bit strange, was it intentional?

File: app/services/disaggregate_community_use.rb [hunk 2]
```diff
@@ -85,7 +85,7 @@ def community_weights(date, meter)
   def disaggregate_whole_meters(fuel_type)
     meters = @school.underlying_meters(fuel_type)
 
-    community_use_meters     = meters.select(&:has_exclusive_community_use?)
+    meters.select(&:has_exclusive_community_use?)
```
Comment (urbanwide) [inline app/services/disaggregate_community_use.rb:88]: Wondering if this line does anything now as it's return value isn't assigned to anything?

File: app/services/usage/annual_usage_meter_breakdown_service.rb [hunk 3]
```diff
@@ -113,41 +114,41 @@ def end_date
...
+    # This means the code is consistent with the chart, but if we want a service that just
+    # generates a usage breakdown based on the last 12 months, then we'll need to tweak the
+    # code and use aggregate_meter.amr_date.end_date instead
     def chart_date_ranges
-      #create a period calculator, to calculate date ranges
+      # create a period calculator, to calculate date ranges
       period_calculator = PeriodsBase.period_factory(chart_config,
-        @meter_collection, aggregate_meter.amr_data.start_date, aggregate_meter.amr_data.end_date)
-      #calculate the ranges
+                                                     @meter_collection, aggregate_meter.amr_data.start_date, aggregate_meter.amr_data.end_date)
```
Comment (urbanwide) [inline app/services/usage/annual_usage_meter_breakdown_service.rb:134]: This spacing looks a bit large...

**Review by urbanwide — COMMENTED** [review]
Looks good - have made a few inline comments, which could be a result of overzealous 
auto-correct... Also can see that this has a few conflicts - let me know if you want me to review again before merging.

File: app/services/disaggregate_community_use.rb (hunk 2 above)
Comment (ldodds) [reply app/services/disaggregate_community_use.rb]: File has been removed in master as its unused

File: app/services/costs/accounting_costs_service.rb (hunk 1 above)
Comment (ldodds) [reply app/services/costs/accounting_costs_service.rb:92]: This was rubcop autocorrect, I've added the variable back in as `_previous_year_end_date`.

File: app/services/usage/annual_usage_meter_breakdown_service.rb (hunk 3 above)
Comment (ldodds) [reply app/services/usage/annual_usage_meter_breakdown_service.rb:134]: Rubocop wants the arguments aligned!

**Review by urbanwide — COMMENTED** [review]
Looks good @ldodds - a few small comments of the same ilk as before🖕

File: spec/app/services/baseload/statistical_baseload_calculator_spec.rb [hunk 4]
```diff
@@ -17,8 +21,15 @@
     end
 
     context 'with varied consumption' do
-      #8 periods of 0.1, rest are random
-      let(:kwh_data_x48)    { Array.new(10, rand(1.1..3.0)) + Array.new(4, 0.1) + Array.new(26, rand(1.1..3.0)) + Array.new(4, 0.1) + Array.new(4, rand(1.1..3.0)) }
+      # 8 periods of 0.1, rest are random
+      let(:kwh_data_x48) do
+        Array.new(10,
+                  rand(1.1..3.0)) + Array.new(4,
```
[the remark this answers is not in the document]
Comment (ldodds) [reply spec/app/services/baseload/statistical_baseload_calculator_spec.rb:27]: I've reworked this. It allows it to be on separate lines, but still indented. So not sure what it doesn't do that by default!

File: spec/app/services/heating/heating_thermostatic_analysis_service_spec.rb [hunk 5]
```diff
@@ -11,22 +11,23 @@
     @beta_academy = load_unvalidated_meter_collection(school: 'beta-academy')
   end
 
-  context '#enough_data?' do
-    let(:service) { Heating::HeatingThermostaticAnalysisService.new(meter_collection: @acme_academy, fuel_type: :gas) }
+  describe '#enough_data?' do
+    let(:service) { described_class.new(meter_collection: @acme_academy, fuel_type: :gas) }
```
[the remark this answers is not in the document]
Comment (ldodds) [reply spec/app/services/heating/heating_thermostatic_analysis_service_spec.rb:15]: I've disabled it now as I agree. But I've left this change in place.

File: spec/app/models/open_close_times_spec.rb [hunk 6]
```diff
@@ -1,31 +1,34 @@
...
-        expect(open_close_times.usage(day)[:school_day_open]).to eq([TimeOfDay.new(7,30)..TimeOfDay.new(16,20)])
+        expect(open_close_times.usage(day)[:school_day_open]).to eq([TimeOfDay.new(7, 30)..TimeOfDay.new(16, 20)])
       end
     end
+
     describe 'with both' do
-      let(:community_use_times)          do
-        [{day: :monday, usage_type: :community_use, opening_time: TimeOfDay.new(19,00), closing_time: TimeOfDay.new(21,30), calendar_period: :term_times}]
+      let(:community_use_times) do
+        [{ day: :monday, usage_type: :community_use, opening_time: TimeOfDay.new(19, 0o0), closing_time: TimeOfDay.new(21, 30), calendar_period: :term_times }]
```
[the remark this answers is not in the document]
Comment (ldodds) [reply spec/app/models/open_close_times_spec.rb]: Yes apparently! I've reverted it back to just `0`

**Review by urbanwide — APPROVED** [review]

**Pull request merged**